doc/aaa/AAA Restrictions/AAA Restrictions.md

```diff
@@ -0,0 +1,252 @@
+
+### Requirements
+
+-      Limit Login Attempts: Specify the maximum number of unsuccessful login attempts allowed before a user account is temporarily locked. This prevents continuous guessing of credentials.
+-      Control Login Speed: Introduce delays between consecutive login attempts to slow down any automated login attempts, significantly reducing the risk of brute force attacks.
+-      Define Unlock Time: Establish a configurable unlock time that specifies how long a user account remains locked before allowing another login attempt. This control helps manage the access of users who have shown suspicious login behaviors.
+-      TODO: SONIC rest API/nginx?
+
+### Architecture Design 
+
+Arc design diagram\
+![arc_aaa_sonic](restrictions.png)
+
+### High-Level Design 
+
+#### Sonic-cli
+The CLI will update the CONFIG_DB aaa tables with the commands mentioned above & will show the current configuration.
+
+#### PAM, NSLCD, NSS
+The file list below will be modified according to the flows of the feature, more description in flow chapter to support login session limitations and aaa.
+/etc/common-auth-sonic or /etc/common-auth
+/etc/pam_talley2 , /etc/pam_faillock , /etc/pam_faildelay.so
+/etc/nsswitch.conf
+
+#### Hostcfgd
+Hostcfgd – listen to changes in CONFIG_DB in the aaa table, and when the table have a new modification/or init happens it will trigger a callback in hostcfgd handle in AAA class to modify the PAM & NSS configuration files in Linux.
+
+#### Database
+will update the Redis CONFIG_DB to have the necessary fields.
+AAA: {
+  Authentication: {
+    restrictions: {
+      lockout-state: (True/False)
+      lockout-reattempt: (duration_sec)*
+      lockout-attempts: (number>2)
+      fail-delay: (duration_sec)
+    }
+  }
+}
+*      lockout-reattempt: 0 is considered without a limit. will not allow reattempt without resetting the counter.
+
+### Init Flow
+#### Compilation
+No new pkg are added in the build time.
+
+#### Feature Default
+Restrictions are enabled by default.
+In addition, this affects the common-auth-sonic file switching to a duplicative branch to use pam_faildelay.so and pam_faillock.so
+      lockout-state: "True"
+      lockout-reattempt: (15 sec)
+      lockout-attempts: (5 attempts)
+      fail-delay: (0 sec)
+
+#### Dependencies
+
+not relevant
+
+### SAI API 
+
+not relevant
+
+#### Manifest (if the feature is an Application Extension)
+
+not relevant
+
+#### CLI/YANG model Enhancements 
+```yang
+//filename: sonic-system-aaa.yang
+module sonic-system-aaa {
+...
+    container sonic-system-aaa {
+        container AAA {
+            list AAA_LIST {
+                key "type";
+
+                leaf type {
+                    type enumeration {
+                        enum authentication;
+			**enum restrictions;**
+**
+                leaf lockout_state {
```

Comment:
Is this the main knob for enabling or disabling this feature?

Reply:
not exactly as it is independent from fail-delay. 
the lockout_state is regarding the locking of users - to not use the blacklist. the delay on each login is not effected by this

Reply:
Then we should either have something that enables this feature as a whole, or add a build flag on whether we would like these items or not.
I'm more for the cli command, as it is cleaner.